src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1330,6 +1331,32 @@ citus_update_node(PG_FUNCTION_ARGS)
 }
 
 
+Datum
+citus_pause_node(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	int32 nodeId = PG_GETARG_INT32(0);
+	List *placementList = NIL;
+
+	WorkerNode *workerNode = FindNodeAnyClusterByNodeId(nodeId);
+	if (workerNode == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_NO_DATA_FOUND),
+						errmsg("node %u not found", nodeId)));
+	}
+
+
+	if (NodeIsPrimary(workerNode))
+	{
+		placementList = AllShardPlacementsOnNodeGroup(workerNode->groupId);
+		LockShardsInPlacementListMetadata(placementList, AccessExclusiveLock);
```

Comment:
Also, LockShardsInPlacementListMetadata seems to only acquire the locks on the coordinator, we should also acquire the same locks from the worker nodes as query from any node is the default behavior. 

We should also acquire the locks on the workers, you can check `BlockWritesToShardList` function for that

Reply:
I agree this is what we want, but I don't think calling BlockWritesToShardList would be enough. The main usecase of this `citus_pause_node` is calling it during failover, when `citus_update_node` is later used in the same transaction, and while one of the nodes is possibly down. A quick look at the code suggests that BlockWritesToShardList fails if it cannot block writes on one node. We wouldn't want that here. We'd want a best-effort locking on the workers, i.e. if one node is unreachable that's fine.

So while I agree we want this. It definitely feels out of scope for a warmup task. I think even just blocking on the coordinator (i.e. the current behaviour) has benefits enough for Patroni for this function to exist. The improvements of blocking on workers and blocking reads as well, seem like additions that we want. But seem better fit for follow up tasks.

Also, if we would do this in `citus_pause_node`, should we do the same thing in `citus_update_node` then?

Reply:
> I agree this is what we want, but I don't think calling BlockWritesToShardList would be enough.

Sure, I mostly shared to show that we need to acquire the same locks on the workers, not necessarily using this function.


> So while I agree we want this. It definitely feels out of scope for a warmup task

My feeling is that citus_pause_node might not be the best warm-up task anyway :) 

>  think even just blocking on the coordinator (i.e. the current behaviour) has benefits enough for Patroni for this function to exist. 

Well, my fear is that we'll keep building features that are not compatible with query from any node (which we started to do in Citus 12.0), then in a few years, we'd need re-iterate to support query-from-any node with all these known limitations.

Even for Patroni, if you are using query-from-any node now, this won't work well anyway.

> But seem better fit for follow up tasks.

Because of the above, it is probably not a follow-up task but part of the current task.


> Also, if we would do this in citus_pause_node, should we do the same thing in citus_update_node then?

yes, we should. 



Overall, I'm deferring the decisions to @JelteF and @gurkanindibay, but I think at a minimum we should pause the traffic from the workers as well.

Reply:
> Well, my fear is that we'll keep building features that are not compatible with query from any node (which we started to do in Citus 12.0), then in a few years, we'd need re-iterate to support query-from-any node with all these known limitations.
> ...
> I think at a minimum we should pause the traffic from the workers as well.

tl;dr this is a new hook into already existing functionality, not a new feature. So, inheriting the bugs of the current functionality seems fine to me.

So, I don't think the difference is that I don't think of `citus_pause_node` as a new feature. Just as a new way of calling into the already existing node pausing functionality in citus_update_node. So I think it's fine if citus_pause_node inherits the bugs of citus_update_node it's pausing functionality (because indeed it's a bug that it doesn't pause on workers). To make that clearer, I suggested in another comment to reuse and share the actual pausing code of citus_update_node, instead of re-implementing it. 

Hopefully that helps with your worry that we'd need to do additional re-iteratation, because we currently already need to re-iterate the pausing logic of citus_update_node. So adding citus_pause_node now, wont impact the amount of re-iteration we need to do at all.

Reply:
> . To make that clearer, I suggested in another comment to reuse and share the actual pausing code of citus_update_node, instead of re-implementing it.

sounds good, would be nice to add a comment saying that it is not compatible with query from any node

Reply:
Is that comment for citus_pause_node_within_txn function